inference/src/basics.rs

```diff
@@ -209,164 +258,191 @@ impl FOModule {
-                    Err(e) => panic!("error in solver: {e}"),
                 }
 
-                solver.save_tee();
-            }
+                CexResult::Unknown(unknown_reasons.join("\n"))
+            })
+            .collect();
+
+        // Check whether any counterexample has been found
+        if let Some(i) =
+            (0..cex_results.len()).find(|i| matches!(cex_results[*i], CexResult::Cex(_)))
+        {
+            return cex_results.remove(i);
         }
 
-        if separate_cores.iter().any(|core| core.is_none()) {
-            log::debug!("Found unknown.");
-            return TransCexResult::Unknown;
+        // Check whether any query has been canceled
+        if cex_results
+            .iter()
+            .any(|res| matches!(res, CexResult::Canceled))
+        {
+            return CexResult::Canceled;
         }
 
-        let core_sizes = separate_cores
+        // Check whether any query has returned 'unknown'
+        if cex_results
             .iter()
-            .map(|core| core.as_ref().unwrap().len())
+            .any(|res| matches!(res, CexResult::Unknown(_)))
+        {
+            log::debug!("{:>8}ms. Found unknown", start_time.elapsed().as_millis());
+            return CexResult::Unknown(
+                cex_results
+                    .into_iter()
+                    .filter_map(|res| match res {
+                        CexResult::Unknown(reason) => Some(reason),
+                        _ => None,
+                    })
+                    .join("\n"),
+            );
+        }
+
+        // Otherwise, all results must be UNSAT-cores
+        let separate_cores = cex_results
+            .into_iter()
+            .map(|res| match res {
+                CexResult::UnsatCore(core) => core,
+                _ => unreachable!(),
+            })
             .collect_vec();
+        let core_sizes = separate_cores.iter().map(|core| core.len()).collect_vec();
+        let unsat_core: HashSet<usize> = separate_cores.into_iter().flatten().collect();
         log::debug!(
-            "Found UNSAT with {} formulas in core (by transition: {:?})",
+            "{:>8}ms. Found UNSAT with {} formulas in core (by transition: {:?})",
+            start_time.elapsed().as_millis(),
             unsat_core.len(),
             core_sizes
         );
-        TransCexResult::UnsatCore(unsat_core)
+        CexResult::UnsatCore(unsat_core)
     }
 
-    pub fn implication_cex(&self, conf: &SolverConf, hyp: &[Term], t: &Term) -> Option<Model> {
+    pub fn implication_cex(&self, confs: &[&SolverConf], hyp: &[Term], t: &Term) -> CexResult {
         let mut core: Core = if self.gradual {
             Core::new(hyp, HashSet::new(), self.minimal)
         } else {
             Core::new(hyp, (0..hyp.len()).collect(), self.minimal)
         };
 
-        loop {
+        let try_conf = |conf: &SolverConf, core: &mut Core| -> CexResult {
```

Comment:
This is actually a core piece of (potentially) general functionality for working with a "meta solver conf" that consists of a sequence of configurations to try. If you can it would be great to factor this out as much as possible. In this PR I'd definitely want this function to be separated into its own function with documentation. Eventually it would be great if we could somehow wrap `Vec<SolverConf>` in a struct with an API similar to a single `SolverConf` (perhaps `Solver` needs to be a trait for this to work).